### DISCUSSION TITLE
SD_MMC and FabGL ESP32 3.3V Board - 16MB Flash / 4 MB PSRAM not work

### DISCUSSION THREAD

**Discussion by andromeda2911** [fdivitto/FabGL#235] [Q&A]
Hi, 
i received my new board, but
sd_mmc with this board not work.
i don't use fabgl library.

the code below  with the board fabgl work.

```
  SPI.begin(14, 35, 12, 13);
  if (!SD.begin(13)) {
    Serial.println("Card Mount Failed");
    return;
  }
```
the code below SD_MMC not work but work with board VGA32 and TTGO T8:
```
 pinMode(35, INPUT_PULLUP);               <= i have tested pin 35,12,14,13,17,16 but it not work
    if(!SD_MMC.begin("/sdcard", true)){
        Serial.println("Card Mount Failed");
        return;
    }
```
the code below work on board TTGO T8
```
pinMode(2, INPUT_PULLUP);
    if(!SD_MMC.begin("/sdcard",true)){
        Serial.println("Card Mount Failed");
        return;
    }
```
the code below work with the board VGA32 (SD)
 ```
 SPI.begin(14, 2, 15, 13);
  if (!SD.begin(13)) {
    Serial.println("Card Mount Failed");
    return;
  }
```

any idea ?

Thanks.

**Comment by fdivitto** [comment]
does it work with fabgl examples (like the filebrowser)?

**Reply by andromeda2911** [reply]
yes it work well with fabgl libray and FileBrowser.
i use esp32 v1.0.6 on arduino 1.8.19.

error is:
E (27) sdmmc_io: sdmmc_io_reset: unexpected return: 0x108
[E][SD_MMC.cpp:85] begin(): Failed to initialize the card (264). Make sure SD card lines have pull-up resistors in place.

The code:
```
pinMode(35, INPUT_PULLUP);   
    if(!SD_MMC.begin("/sdcard", true)){
        Serial.println("Card Mount Failed");
        return;
    }
```

**Comment by fdivitto** [comment]
On this board SD card is connected to:
MISO = 35
MOSI = 12
CLK = 14
CS = 13

so, SPI.begin(14, 35, 12, 13) should work.
Internally FabGL uses esp_vfs_fat_sdspi_mount() to mount SD, without internal pull-up setting.
I will do further tests shortly...

**Reply by andromeda2911** [reply]
waiting a solution i use this code that work

`#define SD_MMC SD`
  ```
SPI.begin(14, 35, 12, 13);
  // pinMode(13, INPUT_PULLUP);
  //if(!SD_MMC.begin("/sdcard", true)){
  if (!SD_MMC.begin(13)) {
    Serial.println("Card Mount Failed");
    return;
  }
```
Because all my code is based on SD_MMC and not SPI as SD.begin ....

**Comment by fdivitto** [comment]
Sorry, I've never used SD_MMC , anyway it seems you should set signals with setPins method, before begin(). However it doesn't use SPI signals (the unique way supported by FabGL board).

**Comment by fdivitto** [comment]
standing to...
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/sdmmc_host.html
...it seems that MMC data pins cannot be changed, so I guess MMC is not supported.

**Reply by andromeda2911** [reply]
Thank you for your help, it does not matter, I will use SD which works on this card.

**Comment by andromeda2911** [comment]
in the PCB at the level of the SD connector there are 2 resistors of 10k (cms 103), if I add two resistors in the other two empty slots should it work with SD_MMC?
I compared with the VGA32 and the TTGO T8, there are 4 resistors.

**Comment by fdivitto** [comment]
Please could you confirm that following code works on the TTGO VGA32 V1.4?

```
#include "SD_MMC.h"

void setup() {
  Serial.begin(115200); delay(500);
  Serial.println("Reset");

  pinMode(2, INPUT_PULLUP);
  if(!SD_MMC.begin("/sdcard", true)) {
    Serial.println("Card Mount Failed");
  } else  {
    Serial.println("OK");
  }
 
}

void loop() {
}
```

**Reply by andromeda2911** [reply]
it not work on your board, 
this code with TTGO VGA32 1.4 and TTGO T8 1.7
with SdFat libray and sdMMC libray and SD library work.

The problem is with your board (tindie)

the code you are wrtiting not work.
```
Reset
E (537) sdmmc_io: sdmmc_io_reset: unexpected return: 0x108
[E][SD_MMC.cpp:85] begin(): Failed to initialize the card (264). Make sure SD card lines have pull-up resistors in place.
```
I had tested all the possibilities with the pins and different codes.

on the other hand, your card works well with the SdFat and SD library.
but I would have liked to use the MMC library.

**Comment by fdivitto** [comment]
To start, I need to run that code on  **TTGO VGA32 1.4**, but it doesn't work for me (sorry I haven't experience on MMC). I get:

```
E (571) sdmmc_common: sdmmc_init_ocr: send_op_cond (1) returned 0x107
E (572) vfs_fat_sdmmc: sdmmc_card_init failed (0x107).
Card Mount Failed
```

**Comment by andromeda2911** [comment]
you are right, the sd_mmc library does not work on the VGA32 v1.4 card, I have just tested.
I thought it worked.
sd_mmc, sdfat, SD work only for TTGO T8 v1.7.

surely the pins are not connected, you need pullup resistors on some pins of the sd card.

[EDIT]
I believe that to use sd_mmc all the pins of the sd card must be connected.
